applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h

```diff
@@ -0,0 +1,314 @@
+   * @param supersonic_boundary_conditions  A list of variables to fix when
+   *    the mach projected onto the normal of the boundary is above 1.
+   */
+  class KRATOS_API(FLUID_DYNAMICS_APPLICATION) ApplyMachDependentBoundaryConditions: public Process
+  {
+  public:
+        ///@name Type Definitions
+        ///@{
+        //
+        typedef ModelPart::NodeType NodeType;
+        typedef Variable<array_1d<double, 3>> VectorVariable;
+        
+        /**
+         * @brief This class validates and manages a variable to fix
+         * and the value to fix it to.
+         * 
+         * Method ActivateIfInsideTimeInterval must be called at the beginning
+         * of the time step to see if the BC is active or not. This will cause
+         * the following behaviour in the other two public methods:
+         * - Enforce:
+         *     - If t inside interval  -> Fix dof
+         *     - If t outside interval -> Do nothing
+         * - Free
+         *     - Always -> Free dof
+         * 
+         * This allows for checking the time only once at the beginning of the
+         * time-step rather than at every node.
+         */
+        class BoundaryConditionUtility
+        {
+        public:
+            BoundaryConditionUtility(const std::string & variable_name,
+                                     const double Value,
+                                     const IntervalUtility & rIntervalUtility);
+            /**
+             * @brief Checks that the current time is within the interval, and
+             * decides accordingly what function to bind to mEnforceInternal
+             * (either FreeDof or DoNothing).
+             */
+            void ActivateIfInsideTimeInterval(const double time);
+
+            /**
+             * @brief Calls mEnforceInternal. This will either fix the Dof
+             * or do nothing depending on the result of 
+             * ActivateIfInsideTimeInterval.
+             */
+            void Enforce(NodeType & rNode) const;
+
+            /// @brief Frees the Dof.
+            void Free(NodeType & rNode) const;
+
+        private:
+            const Variable<double> * mpVariable;
+            const double mValue; // Value to enforce
+            IntervalUtility mInterval;
+            
+            /* @brief Fixes the Dof managed by this class and sets the value
+             * to the one specified by mValue
+             */
+            static void FixDof(const BoundaryConditionUtility & rUtility, NodeType & rNode);
+
+            /** @brief Does nothing. Used to replace FixDof when t is outside
+             * time interval
+             */
+            static void DoNothing(const BoundaryConditionUtility & rUtility, NodeType & rNode);
+
+            /** @brief This object points to the proper (FixDof|DoNothing)
+             * function according to the time interval.
+             */
+            decltype(FixDof) * mEnforceInternal = & DoNothing;
```

Comment:
I'd use an std::function pointer in here (type doesn't need to be inspected).

Reply:
I tested it and function pointers are considerably faster. See https://github.com/KratosMultiphysics/Kratos/pull/9156#discussion_r717437479